feat(code_review): Handle GitHub re-run check requests

armenzg opened the pull request:
For code reviews, we create a GitHub check. In some cases, the customer may request the check to be re-run (e.g. the task failed or timed out).

This handles the GitHub re-run event by scheduling a task which calls Seer with the original run check ID to re-schedule it (it requires https://github.com/getsentry/seer/pull/4173).

Fixes [CW-18](https://linear.app/getsentry/issue/CW-18/forward-check-re-run-webhook-from-sentry-to-seer).

armenzg committed:
WIP

armenzg committed:
More changes

armenzg committed:
More changes

armenzg committed:
Determine success of forward + others

giovanni-guidini commented on fixtures/github.py:
```diff
@@ -3541,3 +3541,16 @@
   }
 }"""
+
+# Simplified example of a check_run rerequested action event
```
are there docs that list the event format if we want to check the full payload later?

armenzg commented on src/sentry/integrations/github/webhook.py:
```diff
@@ -784,6 +787,34 @@ def _handle(
+
+    @property
+    def event_type(self) -> IntegrationWebhookEventType:
+        return IntegrationWebhookEventType.PULL_REQUEST
```
It could be.

armenzg replied:
I've updated this locally.

giovanni-guidini committed:
handle webhook instead of forwarding

Change the logic to actually handle the webhook payload, extract the info that
is relevant to send to Seer and then send it (to an updated endpoint)
See https://github.com/getsentry/seer/pull/4173 for more details.

giovanni-guidini committed:
possibly fix tests

armenzg committed:
More changes

armenzg committed:
Update tests

armenzg committed:
Consolidate tests

armenzg committed:
Rename module

armenzg committed:
Minor change

armenzg committed:
Consolidate tests

armenzg commented on src/sentry/integrations/github/webhook.py:
```diff
@@ -784,6 +785,32 @@ def _handle(
+
+    @property
+    def event_type(self) -> IntegrationWebhookEventType:
+        return IntegrationWebhookEventType.CHECK_RUN
```
Using now the new `CHECK_RUN` type.

armenzg commented on src/sentry/seer/signed_seer_api.py:
```diff
@@ -53,6 +56,53 @@ def make_signed_seer_api_request(
+    base_url = base_url or settings.SEER_AUTOFIX_URL
+    body = orjson.dumps(payload)
+
+    with metrics.timer(
```
TODO: Think if this metric makes sense or if we want to change it.

armenzg commented on src/sentry/testutils/helpers/github.py:
TODO: Find other tests in the codebase which could benefit from this utility module.

armenzg commented on tests/sentry/seer/code_review/test_webhooks.py:
TODO: Have a closer look to these tests to make sure they're well written.

armenzg committed:
Less tests

armenzg commented on src/sentry/seer/signed_seer_api.py:
```diff
@@ -53,6 +56,53 @@ def make_signed_seer_api_request(
+    """
+    Simple wrapper to POST data to Seer with automatic signing.
+
+    This is a simpler alternative to make_signed_seer_api_request for one-off requests.
```
I will look into it further. I think it was that I did not like the necessity of defining a connection pool in the signature. I will look at more examples and see if I change my mind. Thanks for checking!

armenzg committed:
Fix typing

armenzg committed:
Use make signed seer request

armenzg committed:
Check first

armenzg committed:
More change

armenzg committed:
Type github module

armenzg committed:
Handle improper usage

armenzg committed:
Use pydantic

armenzg committed:
Few more changes

armenzg committed:
Everything works now

armenzg committed:
Remove unused types

armenzg committed:
A few more changes

armenzg commented on src/sentry/seer/code_review/webhooks.py:
```diff
@@ -0,0 +1,123 @@
+    return True
+
+
+def _should_handle_github_check_run_event(organization: Organization, action: str) -> bool:
```
TODO: Add consent removal logic
https://github.com/getsentry/seer/blob/76618daa7badc0e9293f8ec1c33d2e4b12d33af5/src/seer/automation/codegen/tasks.py#L74-L84

armenzg commented on src/sentry/seer/code_review/webhooks.py:
```diff
@@ -0,0 +1,123 @@
+        "organization_id": organization.id,
+        "action": validated_event.action,
+        "html_url": validated_event.check_run.html_url,
+        "external_id": validated_event.check_run.external_id,
```
This value is set by Seer when the check is originally created.

See here:
https://github.com/getsentry/seer/blob/0f1291a7bc62afc03dd4b50524956356659744d8/src/seer/automation/codegen/pr_review_step.py#L475-L480

armenzg commented on tests/sentry/seer/code_review/test_webhooks.py:
```diff
@@ -0,0 +1,155 @@
+
+        with self.options({"coding_workflows.code_review.github.check_run.rerun.enabled": True}):
+            with patch("sentry.seer.code_review.webhooks.logger") as mock_logger:
+                self._send_check_run_event(CHECK_RUN_REREQUESTED_ACTION_EVENT_EXAMPLE)
```
Even though Seer is returning 500, we will assert that status code is 204, thus, GitHub will not retry.

armenzg commented on tests/sentry/seer/code_review/test_webhooks.py:
```diff
@@ -0,0 +1,155 @@
+            mock_request.assert_called_once()
+            call_kwargs = mock_request.call_args[1]
+            body = orjson.loads(call_kwargs["body"])
+            assert body == {"original_run_id": str(self.event_dict["check_run"]["external_id"])}
```
We only send the `original_run_id` to Seer.

armenzg committed:
Minor change

armenzg commented on src/sentry/integrations/github/webhook.py:
```diff
@@ -784,6 +785,32 @@ def _handle(
+            return
+
+        # XXX: Add support for registering functions to call
+        handle_github_check_run_event(organization=organization, event=event)
```
I will follow-up in a new PR to make this more generic.

armenzg commented on src/sentry/seer/code_review/webhooks.py:
In the future, I will be making this module more generic to handle other GitHub webhook events. It's good for now.

armenzg committed:
Remaining changes

armenzg committed:
Merge branch 'master' into 12_04/forward_webhook/armenzg

armenzg commented on src/sentry/seer/code_review/tasks.py:
I will generalize this to handle other webhook events in a future PR.

armenzg commented on src/sentry/seer/code_review/tasks.py:
```diff
@@ -0,0 +1,82 @@
+    namespace=seer_code_review_tasks,
+    retry=Retry(times=3, delay=60),
+    silo_mode=SiloMode.REGION,
+    bind=True,  # Bind task instance to access self.request.retries for conditional logging
```
This allows adding `self` to the signature and analyze if we have reached the maximum number of retries so we can log errors then.

armenzg commented on src/sentry/seer/code_review/tasks.py:
```diff
@@ -0,0 +1,82 @@
+        metrics.incr(f"{PREFIX}.outcome", tags={"status": "invalid_task_parameters"})
+        return
+
+    payload = {"original_run_id": original_run_id}
```
This is the only value we need to send to Seer:
https://github.com/getsentry/seer/pull/4173/changes#diff-d022a5605fc7126215740d58900ce6e7c4be1899c6b4809c8e353bda0883a35dR143

armenzg committed:
Actually retry on error

armenzg committed:
Only retry some errors

armenzg committed:
Handle metric with finally

armenzg commented:
@sentry review

armenzg committed:
Track e2e latency

armenzg committed:
Handle other errors

armenzg committed:
Rename + address feedback

armenzg committed:
Simplify eligibility check

armenzg committed:
Last touches

armenzg commented on src/sentry/seer/code_review/webhook_task.py:
```diff
@@ -0,0 +1,130 @@
+    Args:
+        self: Task instance (bound via bind=True)
+        organization_id: The organization ID (for logging/metrics)
+        enqueued_at_str: The timestamp when the task was enqueued
```
This will be used to track how long does an event take to be processed even if the task needs to be retried.

armenzg committed:
Minor change

armenzg commented on src/sentry/seer/code_review/webhook_task.py:
```diff
@@ -0,0 +1,130 @@
+        metrics.incr(f"{PREFIX}.outcome", tags={"status": "invalid_task_parameters"})
+        raise ValueError("Missing required task parameters.")
+
+    is_last_attempt = self.request.retries >= MAX_RETRIES - 1
```
I'm renaming this variable locally. I will push in a moment.

JoshFerge commented on src/sentry/seer/code_review/webhook_task.py:
```diff
@@ -0,0 +1,130 @@
+            "%s.invalid_timestamp",
+            PREFIX,
+            extra={"enqueued_at": enqueued_at_str, "error": str(e)},
+        )
```
i'm kind of curious if this would ever actually fail. do we really need to try/except this?

JoshFerge replied:
and if so, logging the error doesn't seem ideal, could we use sentry issues instead?

JoshFerge commented on src/sentry/seer/code_review/webhook_task.py:
```diff
@@ -0,0 +1,130 @@
+    context = {"organization_id": organization_id, **kwargs}
+    if not options.get("coding_workflows.code_review.github.check_run.rerun.enabled"):
+        logger.info("Skipping rerun request because the option is disabled", extra=context)
+        status = "option_disabled"
```
why set status only to return the next line? does this even do anything?

armenzg committed:
Address feedback

JoshFerge commented on src/sentry/seer/code_review/webhook_task.py:
```diff
@@ -0,0 +1,130 @@
+        status = e.__class__.__name__
+        metrics.incr(f"{PREFIX}.outcome", tags={"status": f"error_{status}"})
+        if self.request.retries >= MAX_RETRIES - 1:
+            logger.exception("%s.error", PREFIX, extra=context)
```
additionally here -- since we use sentry, we'll have the full stacktrace in the issue -- do we really have to log the status?

GabeVillalobos reviewed (commented):
Changes on the integration side seem reasonable to me. Thanks for running this by us!

JoshFerge reviewed (commented):
seems okay. i feel that the error handling is very over-defensive and will result in a lot of noisy telemetry / makes the code more complicated that it needs to be. And more of a stylistic thing but I also feel that a lot of the comments in this PR feel unnecessary as they just restate what the code is doing.

armenzg committed:
Remove logger.exception()

armenzg committed:
Adjust tests

armenzg committed:
Reduce obvious comments + update them

armenzg committed:
Remove unnecessary metric

armenzg committed:
Stricter tests

armenzg committed:
Remove the need for bind

armenzg committed:
Fix

armenzg commented on src/sentry/seer/code_review/webhook_task.py:
```diff
@@ -0,0 +1,124 @@
+
+# This needs to match the value defined in the Seer API:
+# https://github.com/getsentry/seer/blob/main/src/seer/routes/codegen.py
+SEER_PR_REVIEW_RERUN_PATH = "/v1/automation/codegen/pr-review/rerun"
```
This matches https://github.com/getsentry/seer/pull/4173/files#diff-d022a5605fc7126215740d58900ce6e7c4be1899c6b4809c8e353bda0883a35dR132

armenzg commented:
@seer review

JoshFerge reviewed (approved):
changes lgtm! thank you @armenzg!

suejung-sentry commented on src/sentry/seer/code_review/webhooks.py:
```diff
@@ -0,0 +1,97 @@
+    """
+    Determine if the GitHub check_run event should be handled.
+    """
+    if not features.has("organizations:code-review-beta", organization=organization):
```
Since we plan to ship this PR this week, let's use the logic in this [check](https://github.com/getsentry/sentry/blob/8e2d2ccf0313e45cc8c8eeb3db9e58db9dff3138/src/sentry/utils/seer.py#L6)  `can_use_prevent_ai_features`

PLUS

If the org has `seat-based-seer-enabled` flag on (representing that they've bought the new seat-based seer plan), we should ALSO CHECK whether their repo is enabled using something like below:
```
        repo_settings = (
            RepositorySettings.objects.select_related("repository")
            .filter(
                repository__external_id=external_repo_id,
                repository__organization_id=sentry_org_id,
                repository__provider=provider, // "integrations:github"
                repository__status=ObjectStatus.ACTIVE,
                enabled_code_review=True,
            )
            .first()
        )
```

This is the logic I hope to get pulled into an all-in-one helper with the overwatch migration project.

If we want to tackle this complexity separately, I'm good with merging this PR so it just unlocks this for code-review-beta cohort to start. If we want to go this route, let's make sure to also throw in [these](https://github.com/getsentry/sentry/blob/8e2d2ccf0313e45cc8c8eeb3db9e58db9dff3138/src/sentry/utils/seer.py#L8C5-L13C21) legal-required kill switch checks

suejung-sentry reviewed (approved):
This is great! Added a comment on options to update the "code review eligibility" check logic to match the current state of things as we get ready for GA
